lib/postprocessing/watermarkProcessor.js

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2020 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+'use strict';
+
+const { getDimensions } = require('./assetProperties');
+const { ImageOperations } = require('./imageOperations');
+
+class WatermarkProcessor {
+    constructor(instructions) {
+
+        this.renditionWidth = instructions.width || this.getRenditionDimensions(instructions.path).width || 0;
+        this.renditionHeight = instructions.height || this.getRenditionDimensions(instructions.path).height || 0;
+        this.watermarkWidthPercent = instructions.watermark.widthPercent || 100;
+        this.watermarkPath = instructions.watermark.path;
+        this.getWatermarkDimensions(this.watermarkPath);
+    }
+
+    getRenditionDimensions (path) {
+        return getDimensions(path);
+    }
+
+    getWatermarkDimensions (path) {
+        const dimensions = getDimensions(path);
+        this.watermarkWidth = dimensions.width;
+        this.watermarkHeight = dimensions.height;
+    }
+
+    compareRenditionWidth () {
+
+        if( this.renditionWidth <= this.watermarkWidth ) {
+            return this.renditionWidth;
+        }
+        return 0;
+    }
+
+    compareRenditionHeight () {
+
+        if (this.renditionHeight <= this.watermarkHeight) {
+            return this.renditionHeight;
+        }
+        return 0;
+    }
+
+    async fitWatermark(destinationAsset){
+
+        const resizedAsset = destinationAsset || this.watermarkPath.replace("watermark", "watermark-resized");
+        const img = new ImageOperations(this.watermarkPath);
+
+        this.getWatermarkDimensions(this.watermarkPath);
+
+        const watermarkWidthResize = this.compareRenditionWidth();
+
+        if (watermarkWidthResize > 0) {
+            img.resize(watermarkWidthResize);
+        }
+
+        await img.write(resizedAsset);
```

Comment:
Why do we already write the `img` here and later below resize and write it again? And then potentially a 3rd time again. Just for the watermark resizing? Each write is one `imagemagick` invocation...

AFAICS we need to
1. identify the sizes of both rendition and watermark
2. then together with widthPercentage and placement options we know exactly how to size and place the watermark
3. load image and do a single resize
4. on the main rendition `gm` do the composite with that resized watermark file

This should be doable in one function and doesn't really need a separate class such as `WatermarkProcessor` here.

Reply:
And we might be able to do this with a single `gm()` and single imagemagick process invocation:

https://stackoverflow.com/a/40940603/2709

just the first part of the answer leveraging `draw()`:

```
gm(_image)
    // WATERMARK - PARAM ORDER: [X Pos, Y Pos, width, height]
    .draw(['gravity SouthEast image Over 0,0 256,256 "/path/to/watermark.png"'])
    // RESIZE DIMENSIONS - PARAM ORDER: [width, height]
    .resize(2048, null)
    .write("/path/to/resized/output.jpg" ...)
```